### PR TITLE
Java.Util.HashMap

### PR DESCRIPTION
**Pull request by ghost**
Edited + checked old source.
Built + Add test for new method.

### CODE REVIEW

File: java/util/HashMap/HashMapTest.cpp [hunk 1]
```diff
@@ -51,95 +53,111 @@ TEST (JavaUtil, HashMapConstructor) {
...
-	expectedValue = "-44444";
-	result = anotherMap.get(expectedKey);
-	ASSERT_STR(expectedValue.toString(), result.toString());
-
-	// Given invalid key and make result is null
-	String notExpectedKey = "wrong key";
-	result = anotherMap.get(notExpectedKey);
+	hashMap.put("Key1", "Value of Key1");
+	hashMap.put("Key2", "Value of Key2");
+	HashMap<String, String> anotherMap = (HashMap<String, String>) hashMap.clone();
```
Comment (TuanAnhNguyen69) [inline java/util/HashMap/HashMapTest.cpp]: Please remove this cast, is't not necessary

File: java/util/HashMap/HashMapTest.cpp [hunk 2]
```diff
@@ -153,302 +171,299 @@ TEST (JavaUtil, HashMapEntrySet) {
...
+    // Test non-existent key
+    actualKey = "Non-existent key";
+    actualReplace = hashMap.replace(actualKey, newValue);
+    ASSERT_STR("", actualReplace.toString());
+    
 }
 
-TEST (JavaUtil, HashMapReplaceSpecifiedValue) {
-	// Given valid hash map to test
+TEST(JavaUtil, HashMapReplace2) {
```
Comment (TuanAnhNguyen69) [inline java/util/HashMap/HashMapTest.cpp]: Please rename this test

File: kernel/type.h [hunk 3]
```diff
@@ -31,6 +31,7 @@
 #define FALSE 0
 #define NOT_FOUND -1
 #define MAX_STRING_LENGTH 100000
+#define BiConsumer std::function
```
Comment (TuanAnhNguyen69) [inline kernel/type.h]: I think we should comment this and wait for BiConsumer class

File: java/util/HashMap/HashMap.hpp [hunk 4]
```diff
@@ -24,502 +24,760 @@
...
+
+                return true;
+			}
+
+			/**
+			 * forEach method: loop a HashMap and do the function action
+			 *
+			 * @param action
+			 */
+			void forEach (BiConsumer<void(Key, Value)> action) {
```
Comment (TuanAnhNguyen69) [inline java/util/HashMap/HashMap.hpp]: I think we should comment this and wait for BiConsumer class

File: java/util/HashMap/HashMap.hpp [hunk 5]
```diff
@@ -24,502 +24,760 @@
...
+			 * Iterator.remove, Set.remove,
+			 * removeAll, retainAll, and clear
+			 * operations.
+			 *
+			 * It does not support the add or addAll
+			 * operations.
+			 *
+			 * @return a set view of the keys contained in this map
+			 */
+//			Set<Key> keySet() {
```
Comment (TuanAnhNguyen69) [inline java/util/HashMap/HashMap.hpp]: Set class is available, what do you really need in Set class?

**Review by TuanAnhNguyen69 — CHANGES_REQUESTED** [review]
Please check for some mistakes in comments

File: kernel/type.h (hunk 3 above)
Comment (loint) [reply kernel/type.h]: Please remove this definition because it's wrong. Should be BiConsumer implementation

**Review by loint — CHANGES_REQUESTED** [review]
Please remove all binaries in your PR.

File: java/util/HashMap/HashMapTest.cpp (hunk 1 above)
Comment (ghost) [reply java/util/HashMap/HashMapTest.cpp]: Thanks. Fixed.

File: java/util/HashMap/HashMapTest.cpp (hunk 2 above)
Comment (ghost) [reply java/util/HashMap/HashMapTest.cpp]: Thanks. Fixed.

File: kernel/type.h (hunk 3 above)
Comment (ghost) [reply kernel/type.h]: Thanks. Fixed.

File: java/util/HashMap/HashMap.hpp (hunk 4 above)
Comment (ghost) [reply java/util/HashMap/HashMap.hpp]: Thanks. Fixed.

File: kernel/Type.hpp [hunk 6]
```diff
@@ -29,6 +29,7 @@
 
 #define NOT_FOUND -1
 #define MAX_STRING_LENGTH 100000
+#define BiConsumer std::function
```
Comment (dthongvl) [inline kernel/Type.hpp]: Please remove this

File: java/util/HashMap/HashMap.hpp [hunk 7]
```diff
@@ -24,504 +24,749 @@
...
+            ConstIteratorType end() const {
+                return this->original.end();
+            }
+
+        public:
+            /**
+             * Compares the specified object
+             * with this map for equality.
+             *
+             * @param object
```
Comment (dthongvl) [inline java/util/HashMap/HashMap.hpp]: Please update param

File: java/util/HashMap/HashMap.hpp [hunk 8]
```diff
@@ -24,504 +24,749 @@
...
+
+                return true;
+            }
+
+            /**
+             * forEach method: loop a HashMap and do the function action
+             *
+             * @param action
+             */
+            void forEach(BiConsumer<void(Key, Value)> action) {
```
Comment (dthongvl) [inline java/util/HashMap/HashMap.hpp]: Please help me disable this function until BiConsumer is implemented

File: java/util/HashMap/HashMap.hpp [hunk 9]
```diff
@@ -24,504 +24,749 @@
...
+             * entrySet view.  This ensures that m1.equals(m2)
+             * implies that m1.hashCode() == m2.hashCode() for any two maps
+             * m1 and m2, as required by the general contract of
+             * Object.
+             *
+             * This implementation iterates over entrySet(), calling
+             * {Map.Entry hashCode()} on each element (entry) in the
+             * set, and adding up the results.
+             *
+             * @return 	the hash code value for this map
```
Comment (dthongvl) [inline java/util/HashMap/HashMap.hpp]: int

File: java/util/HashMap/HashMap.hpp [hunk 10]
```diff
@@ -24,504 +24,749 @@
...
+            /**
+             * Returns the hash code value for this map.  The hash code of a map is
+             * defined to be the sum of the hash codes of each entry in the map's
+             * entrySet view.  This ensures that m1.equals(m2)
+             * implies that m1.hashCode() == m2.hashCode() for any two maps
+             * m1 and m2, as required by the general contract of
+             * Object.
+             *
+             * This implementation iterates over entrySet(), calling
+             * {Map.Entry hashCode()} on each element (entry) in the
```
Comment (dthongvl) [inline java/util/HashMap/HashMap.hpp]: Please remove { }

File: java/util/HashMap/HashMap.hpp [hunk 11]
```diff
@@ -24,504 +24,749 @@
...
+                    }
+                }
+                return false;
+            }
+
+            /**
+             *  Set of hash map entries
+             *
+             * Returns a Set view of the mappings contained in this map.
+             * @return Set<Map.Entry<Key,Value>> - a set view of the mappings contained in this map
```
Comment (dthongvl) [inline java/util/HashMap/HashMap.hpp]: Please help me update return data type.

File: java/util/HashMap/HashMap.hpp [hunk 12]
```diff
@@ -24,504 +24,749 @@
...
+                    Map<String, String>::Entry entry(pair.first, pair.second);
+                    entrySet.add(entry);
+                }
+
+                return entrySet;
+            }
+
+            /**
+             * Returns the value to which the specified key is mapped,
+             * or null if this map contains no mapping for the key
```
Comment (dthongvl) [inline java/util/HashMap/HashMap.hpp]: default constructor or something, not null

File: java/util/HashMap/HashMapTest.cpp [hunk 13]
```diff
@@ -33,444 +33,835 @@ using namespace Java::Lang;
...
-		if (counter == 2) {
-			assertEquals("Key 98", entry.getKey().toString());
-			assertEquals("Value 98", entry.getValue().toString());
-		}
-	}
-	// Make sure foreach is working
-	assertEquals(100, counter);
+    HashMap<String, String> hashMap;
+
+    for (int index = 1; index <= 100; index++) {
```
Comment (dthongvl) [inline java/util/HashMap/HashMapTest.cpp]: Please create `int index;` outside the loop

**Review by dthongvl — CHANGES_REQUESTED** [review]
Please remove unused files and ignore them.
Please update methods' s comment that use null

File: java/util/HashMap/HashMap.hpp (hunk 10 above)
Comment (ghost) [reply java/util/HashMap/HashMap.hpp]: Thanks. Fixed.

File: java/util/HashMap/HashMap.hpp (hunk 8 above)
Comment (ghost) [reply java/util/HashMap/HashMap.hpp]: Thanks. Fixed.

File: java/util/HashMap/HashMap.hpp (hunk 7 above)
Comment (ghost) [reply java/util/HashMap/HashMap.hpp]: Thanks. Fixed.

File: java/util/HashMap/HashMap.hpp (hunk 9 above)
Comment (ghost) [reply java/util/HashMap/HashMap.hpp]: Thanks. Fixed.

File: java/util/HashMap/HashMap.hpp (hunk 11 above)
Comment (ghost) [reply java/util/HashMap/HashMap.hpp]: Thanks. Fixed.

File: java/util/HashMap/HashMap.hpp (hunk 12 above)
Comment (ghost) [reply java/util/HashMap/HashMap.hpp]: Thanks. Fixed.

File: java/util/HashMap/HashMapTest.cpp (hunk 13 above)
Comment (ghost) [reply java/util/HashMap/HashMapTest.cpp]: Thanks. Fixed.

File: kernel/Type.hpp (hunk 6 above)
Comment (ghost) [reply kernel/Type.hpp]: Thanks. Fixed.

**Comment by tiny-express** [comment]
This PR will be close because no progress.

**Pull request merged**